### PR TITLE
XRT-2915: add iot_data_restrict_element 

### PR DESCRIPTION
**Pull request by dpazj**

### CODE REVIEW

**Comment by SteveOss** [comment]
> @SteveOss not sure what we want the behaviour to be when the list/vector is empty? -> currently I am just returning IOT_DATA_MULTI

Makes sense to me

**Review by SteveOss — COMMENTED** [review]
Is there any advantage to having multiple iot_data_TYPE_restrict functions as opposed to a single polymorphic iot_data_restrict function ?

File: src/c/data.c [hunk 1]
```diff
@@ -2942,6 +2895,43 @@ extern iot_data_t * iot_data_update_at (const iot_data_t * data, const iot_data_
   return result;
 }
 
+iot_data_type_t iot_data_restricted_element_type (const iot_data_t * data)
+{
+  assert (data && (data->type == IOT_DATA_VECTOR || data->type == IOT_DATA_LIST || data->type == IOT_DATA_ARRAY || data->type == IOT_DATA_MAP));
+  iot_data_type_t restricted_type = data->element_type;
+  if (restricted_type != IOT_DATA_MULTI) goto done;
```
Comment (SteveOss) [inline src/c/data.c:2902]: Should this be:

 if (restricted_type != IOT_DATA_MULTI || data->type == IOT_DATA_ARRAY) goto done;

Comment (dpazj) [reply src/c/data.c:2902]: Yep - I was under the assumption that an IOT_DATA_ARRAY's element_type would never be set to multi. Better to not assume though.

Comment (SteveOss) [reply src/c/data.c]: There is an assert in iot_data_alloc_array that type <= IOT_DATA_BOOL, so can't happen

Comment (SteveOss) [reply src/c/data.c]: The array type maps to C array, so polymorphism is impossible (hence vector)

**Review by SteveOss — APPROVED** [review]
LGTM. For completeness would add utests for ensuring the alloc_typed functions return the correct type.

**Comment by dpazj** [comment]
@SteveOss I have added some extra unit tests using alloc_typed functions - is that what you were looking for?

**Comment by SteveOss** [comment]
> @SteveOss I have added some extra unit tests using alloc_typed functions - is that what you were looking for?

Yep.

**Comment by SteveOss** [comment]
@dpazj Looks like Jenkins issue, ok for me to merge ? Think needs a release note update first.

**Comment by dpazj** [comment]
@SteveOss just added release note update. Everything good to merge from my end.

**Comment by SteveOss** [comment]
@dpazj Will rebuild devel and release packages ...

**Pull request merged**